### ISSUE TITLE
Partial fit / Incremental UMAP

### ISSUE THREAD

**Issue by parashardhapola** [lmcinnes/umap#62]
Hi @lmcinnes,

There has been quite an exciting about UMAP in scRNA-Seq area recently. However, fitting larger data-sets might be an issue depending on amount of available memory. I would love to see  a 'partial_fit' method (akin to one available in scikit's incremental PCA) in UMAP, so that data could be lazily loaded and fitted. I do realize that this might be non-trivial, if not theoretically infeasible, but your recent implementation of 'transform' method got my hopes high. It would be nice to have your thoughts on this.

Thanks.

**Comment by lmcinnes** [comment]
If I understand you correctly you want essentially an "out-of-core" fit method that doesn't require the whole dataset to be in memory at one time?

I believe that this is possible, but would likely be slow. Essentially what is really needed is a fast, out-of-core, approximate nearest neighbor algorithm. I believe what I have now may actually work with [dask arrays](http://dask.pydata.org/en/latest/) which would provide out-of-core, but for the fact that ``check_array`` converts it to an array at the start, and modulo how numba handles dask (I suspect it doesn't). In other-words I think this is algorithmically quite tractable, but might require a fair amount of code to actually make it happen.

If there is significant interest in this (and it seems there may be) I'll see if I can figure out how to actually make this happen.

**Comment by parashardhapola** [comment]
Yes indeed, I'm looking for an out-of-core fit method and I'm sure that it would be of interest to many. 

Pardon me if this is a very naive suggestion but I was thinking that may be this functionality can initially be implemented downstream of distance calculation. Such that, in effect, a user needs to provide a generator which yields distance vector for each sample (of shape m for an m x m distance matrix), instead of complete pre-computed matrix itself. This may avoid, for now, the requirement of an out-of-core ANN algorithm.

**Comment by stsievert** [comment]
> Essentially what is really needed is a fast, out-of-core, approximate nearest neighbor algorithm.

Is this what we were working on during the SciPy sprints on some branch/fork https://github.com/lmcinnes/pynndescent? I'd like to see the code – I'm curious what progress has been made, and would be happy to help out more.

**Comment by lmcinnes** [comment]
Yes, that's the one. I got basic code hammered out, but hadn't fixed all the bugs. I also chatted with Matt Rocklin and it looks like the proposed solution would be quite slow due to the random(ish) access of a large distributed array required. Still, slow but tractable is better than impossible.

**Comment by stsievert** [comment]
> it looks like the proposed solution would be quite slow due to the random(ish) access of a large distributed array required.

I believe that. I sounds like the proposed solution will be enabled by https://github.com/dask/dask/issues/3409.

Is there anywhere I can review the code you've implemented?

**Comment by lmcinnes** [comment]
I didn't get it working to a point of pushing it up anywhere. I have a
colleague who may be picking up this work, so I might try and loop him in
with all of this on Monday.

**Comment by mrocklin** [comment]
On the dask side I think that https://github.com/dask/dask/issues/3409#issuecomment-405254656 is probably the next step to try (assuming I understand the situation on the UMAP side (which is doubtful)) :)

**Comment by soupault** [comment]
Really looking forward to it! Out-of-core implementation would benefit computer vision domain quite a lot.

**Comment by ksangeeta2429** [comment]
Was there something already released for this? I am also looking for the same for acoustic deep learning purpose.

**Comment by lmcinnes** [comment]
Currently this is still, at best, an "in the works" issue. The first real hurdle to be overcome, the nearest neighbor search, is looking good. There is a paralllel NN search, and there is reasonable hope to make a few more changes to ensure it is dask enabled for distributed arrays -- this has all been put the the [PyNNDescent](https://github.com/lmcinnes/pynndescent) library, which will become a dependency for umap in the future. The next step is distributed SGD. This is easier, but has had less work done. Speculatively I would suggest that these are features that may arrive in version 0.5. Note, of course, that 0.4 is not even released yet. I suspect that means you may have to use subsampling for your project now. In good news the structures found my umap are generally fairly robust to subsampling.

**Comment by stsievert** [comment]
> next step is distributed SGD.

https://github.com/dask/dask/issues/3409 is related (and mentioned above). It kept popping up in my Dask distributed SGD implementation (https://github.com/dask/dask-glm/pull/69).

https://github.com/dask/dask/pull/3901 resolves this issue (thanks to @TomAugspurger). Now, shuffling to a distributed Dask array is much quicker (~100x improvement, 13.7s vs 140ms).

**Comment by teenoc** [comment]
would really like to see the incremental training feature for UMAP. I work with both image and text data, UMAP easily generates the best manifold with the least amount of tuning compared to everything else I've tried in both categories.

**Comment by lmcinnes** [comment]
At present it looks like this is something that might be possible in the v0.5 to v0.6 time-frame. That's not soon, but it may yet happen.

**Comment by natemccoy** [comment]
I am writing to mention that I am very interested in this.

**Comment by jamesthegiantpeach** [comment]
A possible method to do a large scale nearest neighbor search efficiently would be to use something like k-d trees or something like (this)[https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/32616.pdf]. I had attempted this a while ago and found that it works for lower dimensional data like vgg16. I tried to use it on some very high dimensional genomics applications and had no luck. 

I’ll try to dig up this code or my notes on this.

**Comment by jamesthegiantpeach** [comment]
I think the solution here will almost certainly involve coming up with a good spatial tree for indexing the inputs so they can be partitioned into blocks. I would be interested in working on this if I found something that could index a large number of points in R^n for large n.

**Comment by galinaalperovich** [comment]
Would be happy to this incremental UMAP feature!

**Comment by lmcinnes** [comment]
Depending on what exactly you are looking for, this may now be available (in preliminary form) in the 0.5dev branch. To make it work you would use the new, special, AlignedUMAP class, and the ``update`` functionality therein. I have a [notebook gist](https://gist.github.com/lmcinnes/a655c9382401a3e07915310f7611e8cf) that provides a basic walkthrough of these sorts of features. With this approach if your goal is to incrementally add (batches) of new points to an existing embedding, updating the embedding as you go, this should suffice. It may not meet all of your needs, but hopefully it will solve some problems for some users.

**Comment by fedden** [comment]
Thanks for the early release and consideration. Just wondering if there is an approach to take wrt the relation dictionaries if the successive instances in the dataset _do not_ depend on each other (no overlap) and as such do not have relations between successive frames.

**Comment by lmcinnes** [comment]
That's definitely a bug -- as you say, it should treat that as no relation. I presume you have a dataset you want aligned except at some intermediate slice where you lack relations? I'll have to dig in a little and see if this is easy to fix. Having at least some relation is baked in to the assumptions right now, but I can see that having empty relations occasionally does make sense.

**Comment by irl-dan** [comment]
Did this ever make it into any released version of the package? Or any plans on doing so? I'm also interested in this.

**Comment by jmdelahanty** [comment]
I too have an application potentially that would be very helpful to have this for!

**Comment by AtomicCactus** [comment]
Just want to +1 this request, an incremental out-of-core fit method would be awesome!

**Comment by SbstnErhrdt** [comment]
Joining +1

**Comment by JdbermeoUZH** [comment]
+1!

**Comment by ECburx** [comment]
+1

**Comment by mahabenfares** [comment]
Hello, I want to use UMAP for data streams. Do you offer an incremental version of UMAP that I can use please? I saw that you talked about Alignedumap, I wanted to test it, but I don't know what to put in the relation dictionaries.

**Comment by lmcinnes** [comment]
I think a lot of this is beyond what the basic UMAP approach can do as it is essentially transductive. For those who really need this I would suggest your best option is going to be using the ParametricUMAP option and using new batches to retrain. If you need pure streaming you can use the transform on ParametricUMAP for the streaming data, and store batches for batch based retraining on a cycle that is reasonable given the retrain cost.

**Comment by big-o** [comment]
I'm not sure I fully understand all of the issues, but is a scalable nearest neighbour search still a blocker? If so, I've had good successes with the [FAISS](https://github.com/facebookresearch/faiss) implementation of an approximate NN search (went from days with the standard ball tree stuff to seconds), and it can be [wrapped up fairly trivially](https://towardsdatascience.com/make-knn-300-times-faster-than-scikit-learns-in-20-lines-5e29d74e76bb) in a sklearn API. The downside is that it limits you to Euclidean and Cosine distance measures only - not sure if this limitation is due to the algorithm or because other distances just haven't been implemented yet.

**Comment by lagunabravo** [comment]
Thank you @lmcinnes for your great work. 
I’m uncertain about how to use retraining on ParametricUMAP as you suggested on Mar 4. I guess what I don’t understand is how retraining modifies the embedder in regard to different batches on a large dataset. Should I split the dataset into batches, train/retrain across all of them and them use the final embedder to apply transform across the whole dataset? Or, alternatively, should I apply transform after retraining for each individual batch?

**Comment by AlekseyFedorovich** [comment]
I would love this feature!!
